Fix IndexFactory.writeIndex for tabix indexes

magicDGS opened the pull request.

yfarjoun commented on src/main/java/htsjdk/tribble/index/IndexFactory.java:
```diff
@@ -292,7 +294,9 @@ public static LinearIndex createLinearIndex(final File inputFile, final FeatureC
         try {
-            stream = new LittleEndianOutputStream(new BufferedOutputStream(new FileOutputStream(idxFile)));
+            final OutputStream underlyingStream = (idx instanceof TabixIndex) ?
+                    new BlockCompressedOutputStream(idxFile) : new FileOutputStream(idxFile);
```
would it be possible for the index to know what type of outputstream it needs and either open it for the factory, or write itself to that stream (given a file). it seems that it's more appropriate to give that decision to the index rather than to the factory.

also: tests please.

magicDGS replied:
I was thinking about this before, because actually it makes more sense to have a `Index.write(File idxFile)` method instead of the factory. But I decided to go for this to maintain compatibility. In addition, the javadoc said that "It is the responsibility of this class to determine and create the correct index type from the input file or stream.", so that's why I did it like this.

magicDGS commented:
@yfarjoun, I don't know how to do tests for checking this...

yfarjoun commented on src/main/java/htsjdk/tribble/index/Index.java:
```diff
@@ -26,6 +26,7 @@
 import htsjdk.tribble.util.LittleEndianOutputStream;
 
 import java.io.File;
+import java.io.FileNotFoundException;
```
where is this used?

yfarjoun commented on src/main/java/htsjdk/tribble/index/AbstractIndex.java:
```diff
@@ -25,6 +25,7 @@
 
 import java.io.BufferedOutputStream;
 import java.io.File;
+import java.io.FileNotFoundException;
```
doesn't seem to be used

yfarjoun commented:
couldn't you test that you can write to a file (and an index is created) and write to a stream (and no exception is thrown)?

magicDGS commented:
@yfarjoun, I added a test to assert that `write(File idxFile)` is writing a file, and that `write( LittleEndianOutputStream stream)` to a null stream is not throwing an error. This was done for both `LinearIndex` and `TabixIndex`. Are they enough?

magicDGS commented:
I also squashed the commits. I think that now that the writing to the file is inside the index, the method `IndexFatory.writeIndex(final Index idx, final File idxFile)` could be deprecated. If someone already have the index file, `idx.write(idxFile)` could be used...

magicDGS commented:
Can you have a look to this, @yfarjoun? It is important for other PR that I'm preparing...

magicDGS commented:
Could you have a look, @yfarjoun? Thanks in advance!

yfarjoun commented on src/main/java/htsjdk/tribble/index/AbstractIndex.java:
```diff
@@ -343,13 +343,20 @@ public void write(final LittleEndianOutputStream stream) throws IOException {
-                new LittleEndianOutputStream(new BufferedOutputStream(new FileOutputStream(Tribble.indexFile(featureFile))));
-        write(idxStream);
-        idxStream.close();
+                new LittleEndianOutputStream(new BufferedOutputStream(new FileOutputStream(idxFile)));
```
extra indent?

magicDGS replied:
It comes from the previous version, but I've just put everything in a single line.

yfarjoun commented:
@lbergelson could you take a look? we originally reviewed this together.

lbergelson commented on src/test/java/htsjdk/tribble/index/IndexFactoryTest.java:
```diff
@@ -112,4 +117,45 @@ public void testCreateTabixIndexOnBlockCompressed() {
+    @Test
+    public void testWriteIndex() throws Exception {
+        // temp directory for this test
+        final File tempDir = IOUtil.createTempDir("writeIndex", null);
```
I think the standard in htsjdk is to use `File output = File.createTempFile()` to get your output file, and then call `output.deleteOnExit()` to make sure it's cleaned up afterwards.

magicDGS replied:
I change it in the new test.

lbergelson commented on src/test/java/htsjdk/tribble/index/IndexFactoryTest.java:
```diff
@@ -112,4 +117,45 @@ public void testCreateTabixIndexOnBlockCompressed() {
     }
+
+    @Test
+    public void testWriteIndex() throws Exception {
```
Since this test is in indexFactory.writeIndex() it should probably be testing `indexFactory.writeIndex()`...  

Test for `write` on the index files themselves should live in TabixIndexTest and LinearIndexTest respectively. 

Either way (or both might be the best at the risk of redundantly testing things), `TabixIndexTest.readWriteTest` provides a good model for how the test might be made.  Instead of repeating the code for the different cases you can write a test function that takes a dataprovider with several inputs.  It might be a pain though for linear indexes since I don't see a read or equals method on them.

magicDGS replied:
I changed exactly the same method (with a data provider instead) to `IndexTest`, because I think that it could live in that package (it is testing the interface method for different implementations).

I added a check for the sequence names and properties loaded and written, and a test bed file too.

lbergelson replied:
Good decision 👍

lbergelson commented on src/test/java/htsjdk/tribble/index/IndexFactoryTest.java:
```diff
@@ -112,4 +117,45 @@ public void testCreateTabixIndexOnBlockCompressed() {
+        final TabixIndex tabixIndexVcfGz =
+                IndexFactory
+                        .createTabixIndex(inputFileVcfGz, new VCFCodec(), TabixFormat.VCF, null);
+        // wirte to a file
```
typo wirte -> write

magicDGS replied:
Changed in the new test.

magicDGS commented on src/main/java/htsjdk/tribble/index/AbstractIndex.java:
```diff
@@ -343,13 +343,19 @@ public void write(final LittleEndianOutputStream stream) throws IOException {
     }
 
     @Override
+    public void write(final File idxFile) throws IOException {
```
Done

magicDGS commented on src/main/java/htsjdk/tribble/index/IndexFactory.java:
```diff
@@ -290,16 +287,7 @@ public static LinearIndex createLinearIndex(final File inputFile, final FeatureC
      * @throws IOException
      */
     public static void writeIndex(final Index idx, final File idxFile) throws IOException {
```
Done

magicDGS commented on src/main/java/htsjdk/tribble/index/tabix/TabixIndex.java:
```diff
@@ -201,6 +201,7 @@ public TabixFormat getFormatSpec() {
      * @param tabixFile Where to write the index.
      */
+    @Override
     public void write(final File tabixFile) {
```
Done. Now throwing `IOException`

magicDGS commented:
Addressed comments, @lbergelson. I changed the method to throw `IOException`, although I'm waiting for @yfarjoun comments about it.

Now that I'm looking at the tribble package, I that an issue should be open to review exceptions thrown by the classes. It will be useful for distinguish them. For instance, when I implemented `getTabixFormat()` it throws a generic `TribbleException`, but I can create another extension of the class that will be more informative.

magicDGS commented:
Tests are failing for FTP issues, @yfarjoun and @lbergelson...

lbergelson commented:
@magicDGS I restarted the tests.  I hate those ftp tests. 

This gets my approval.  This is a breaking change with the changes to the exceptions, but I think it's a positive one to make them more consistent.  Lets hear what @yfarjoun thinks before merging though. 

I agree with you that the current state of exceptions is pretty random.  It would be good to do an overhaul of them.

yfarjoun commented:
a strange test is failing again. otherwise 👍

lbergelson commented:
Oh yeah, that one.  #578 strikes again. Restarted.

magicDGS commented:
Thanks for review and accept this change, @yfarjoun and @lbergelson. I also opened an issue to have a look to the exceptions thrown in the package.
